Further adjustments to the organization storage access prompt

sysrqb opened the pull request:
#### 1d87ffb4145cbaff9fc2b97b2bd6d7d2f3dda29a
<pre>
Further adjustments to the organization storage access prompt
<a href="https://bugs.webkit.org/show_bug.cgi?id=271007">https://bugs.webkit.org/show_bug.cgi?id=271007</a>
<a href="https://rdar.apple.com/124640260">rdar://124640260</a>

Reviewed by Wenson Hsieh.

After receiving additional feedback, adjust the wording and add a disclosure
list. If the list of domains is more than 2, then show the domains in a list
via a disclosure button.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.h:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(-[_WKSSOSiteList initWithSiteList:withAlert:withTableView:]):
(-[_WKSSOSiteList numberOfRowsInTableView:]):
(-[_WKSSOSiteList tableView:viewForTableColumn:row:]):
(-[_WKSSOSiteList toggleTableViewContents:]):
(WebKit::presentStorageAccessAlert):
(WebKit::presentStorageAccessAlertQuirk):
(WebKit::presentStorageAccessAlertSSOQuirk):
(WebKit::displayStorageAccessAlert):

Canonical link: <a href="https://commits.webkit.org/276435@main">https://commits.webkit.org/276435@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0259671d6e8a30595cb329171a7d9037d8eb525f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43993 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/23061 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46431 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46641 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/40060 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46297 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/26971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/36282 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44571 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/20090 "Passed tests") | [⏳ 🧪 api-mac](https://ews-build.webkit.org/#/builders/API-Tests-macOS-EWS "Waiting to run tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17302 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17584 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38979 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/2051 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40170 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39270 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48219 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18996 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/15568 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/43126 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20376 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41845 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9945 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20588 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/20004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -87,18 +149,33 @@ void presentStorageAccessAlertSSOQuirk(WKWebView *webView, const String& organiz
-
-    auto initialDomainList = makeStringByJoining(uniqueDomainList.span(), ", "_s);
+    if (allDomains.size() < 2)
+        completionHandler(true);
```
```suggestion
        return completionHandler(true);
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+
+- (id)initWithSiteList:(NSArray<NSString *> *)siteList withAlert:(RetainPtr<NSAlert>)alert withTableView:(RetainPtr<NSTableView>)tableView
+{
+    _siteList = [[NSArray alloc] initWithArray:siteList copyItems:YES];
```
```suggestion
    _siteList = adoptNS([[NSArray alloc] initWithArray:siteList copyItems:YES]);
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+@end
+
+@implementation _WKSSOSiteList {
+    NSArray<NSString *> *_siteList;
```
```suggestion
    RetainPtr<NSArray<NSString *>> _siteList;
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+
+- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row
+{
+    NSString *viewIdenfier =[NSString stringWithFormat:@"row%ldIdentifier", (long)row];
```
```suggestion
    NSString *viewIdenfier = [NSString stringWithFormat:@"row%ldIdentifier", (long)row];
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+- (NSView *)tableView:(NSTableView *)tableView viewForTableColumn:(NSTableColumn *)tableColumn row:(NSInteger)row
+{
+    NSString *viewIdenfier =[NSString stringWithFormat:@"row%ldIdentifier", (long)row];
+    NSTextView *result = [tableView makeViewWithIdentifier:viewIdenfier owner:self];
```
```suggestion
    RetainPtr result = [tableView makeViewWithIdentifier:viewIdenfier owner:self];
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+    NSString *viewIdenfier =[NSString stringWithFormat:@"row%ldIdentifier", (long)row];
+    NSTextView *result = [tableView makeViewWithIdentifier:viewIdenfier owner:self];
+    if (result == nil) {
+        result = [[NSTextView alloc] initWithFrame:NSMakeRect(0, 0, 100, 300)];
```
```suggestion
        result = adoptNS([[NSTextView alloc] initWithFrame:NSMakeRect(0, 0, 100, 300)]);
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+    else
+        frame.size.height += _tableView.get().frame.size.height;
+    [_alert.get().window setFrame:frame display:YES];
+    [_alert.get() layout];
```
```suggestion
    [_alert layout];
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -111,6 +188,38 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
     [alert setMessageText:alertTitle];
     [alert setInformativeText:informativeText];
+    if (accessoryTextList.count) {
+        NSButton *disclosureButton = [[NSButton alloc] init];
```
```suggestion
        RetainPtr disclosureButton = adoptNS([[NSButton alloc] init]);
```

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -111,6 +188,38 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
+        siteListTableView.get().enabled = NO;
+        siteListTableView.get().usesSingleLineMode = YES;
+        siteListTableView.get().hidden = YES;
+        [siteListTableView addTableColumn:[[NSTableColumn alloc] initWithIdentifier:@"columnIdentifier"]];
```
Need `adoptNS` here too.

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -111,6 +188,38 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
+        siteListTableView.get().hidden = YES;
+        [siteListTableView addTableColumn:[[NSTableColumn alloc] initWithIdentifier:@"columnIdentifier"]];
+
+        _WKSSOSiteList *ssoSiteList = [[_WKSSOSiteList alloc] initWithSiteList:accessoryTextList withAlert:alert withTableView:siteListTableView];
```
`adoptNS`

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,64 @@
+    RetainPtr<NSTableView> _tableView;
+}
+
+- (id)initWithSiteList:(NSArray<NSString *> *)siteList withAlert:(RetainPtr<NSAlert>)alert withTableView:(RetainPtr<NSTableView>)tableView
```
```suggestion
- (instancetype)initWithSiteList:(NSArray<NSString *> *)siteList withAlert:(RetainPtr<NSAlert>)alert withTableView:(RetainPtr<NSTableView>)tableView
```

whsieh replied:
You'll also want to call into the superclass initializer here.

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -41,6 +41,67 @@
+{
+    NSString *viewIdenfier = [NSString stringWithFormat:@"row%ldIdentifier", (long)row];
+    RetainPtr result = [tableView makeViewWithIdentifier:viewIdenfier owner:self];
+    if (result == nil) {
```
Nit - we generally prefer `!result` instead of checking explicitly against `nil`

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -111,9 +193,42 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
+        [disclosureButton setButtonType:NSButtonTypePushOnPushOff];
+
+        RetainPtr relatedWebsitesLabel = adoptNS([NSTextField labelWithString:accessoryLabel]);
+        RetainPtr disclosureStackView = adoptNS([NSStackView stackViewWithViews:@[disclosureButton.get(), relatedWebsitesLabel.get()]]);
```
```suggestion
        RetainPtr relatedWebsitesLabel = [NSTextField labelWithString:accessoryLabel];
        RetainPtr disclosureStackView = [NSStackView stackViewWithViews:@[disclosureButton.get(), relatedWebsitesLabel.get()]];
```

These are not +1 objects, so you don't want to use `adoptNS` here

whsieh commented on Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
```diff
@@ -111,9 +193,42 @@ void displayStorageAccessAlert(WKWebView *webView, NSString *alertTitle, NSStrin
+        siteListTableView.get().dataSource = ssoSiteList.get();
+        siteListTableView.get().delegate = ssoSiteList.get();
+
+        RetainPtr accessoryStackView = adoptNS([NSStackView stackViewWithViews:@[disclosureStackView.get(), siteListTableView.get()]]);
```
(Ditto — should remove the `adoptNS`)

Commit:
Further adjustments to the organization storage access prompt
https://bugs.webkit.org/show_bug.cgi?id=271007
rdar://124640260

Reviewed by Wenson Hsieh.

After receiving additional feedback, adjust the wording and add a disclosure
list. If the list of domains is more than 2, then show the domains in a list
via a disclosure button.

* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.h:
* Source/WebKit/UIProcess/Cocoa/WKStorageAccessAlert.mm:
(-[_WKSSOSiteList initWithSiteList:withAlert:withTableView:]):
(-[_WKSSOSiteList numberOfRowsInTableView:]):
(-[_WKSSOSiteList tableView:viewForTableColumn:row:]):
(-[_WKSSOSiteList toggleTableViewContents:]):
(WebKit::presentStorageAccessAlert):
(WebKit::presentStorageAccessAlertQuirk):
(WebKit::presentStorageAccessAlertSSOQuirk):
(WebKit::displayStorageAccessAlert):

Canonical link: https://commits.webkit.org/276435@main
